crates/sui-framework/sources/LockedCoin.move

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+module Sui::LockedCoin {
+    use Sui::Balance::{Self, Balance};
+    use Sui::Coin::{Self, Coin};
+    use Std::Errors;
+    use Sui::ID::{Self, VersionedID};
+    use Sui::Transfer;
+    use Sui::TxContext::{Self, TxContext};
+
+    /// The locked_until time passed into the creation of a locked coin is invalid.
+    const EINVALID_LOCK_UNTIL: u64 = 0;
+    /// Attempt is made to unlock a locked coin that has not been unlocked yet.
+    const ECOIN_STILL_LOCKED: u64 = 1;
+
+    /// A coin of type `T` locked until `locked_until_epoch`.
+    struct LockedCoin<phantom T> has key, store {
```

Comment:
Sui Move gotcha: `has store` means that this coin allows unconditional transfers, which we do not want for locked tokens.

Reply:
Good point. I needed `store` here so in `Delegation::undelegate` we can transfer the locked coin back to its original owner. Is the right thing to do here to make `Delegation` a friend of `LockedCoin` and `LockedCoin::transfer` a friend function?

Reply:
`friend` would be one way. Another idea would be to expose `LockedCoin::new(balance: Balance<T>, locked_until_epoch: u64, owner: address)` and then call that from `Delegation`. I think that might be slightly cleaner, since you keep the transfers encapsulated inside `LockedCoin`?